core/src/main/java/org/infinispan/interceptors/impl/CacheWriterInterceptor.java

```diff
@@ -258,11 +259,15 @@ protected void processIterableBatch(InvocationContext ctx, PutMapCommand cmd, Pe
       if (getStatisticsEnabled())
          cacheStores.addAndGet(cmd.getMap().size());
 
-      Iterable<MarshalledEntry> iterable = () -> cmd.getMap().keySet().stream()
+      Iterator<MarshalledEntry> iterator = cmd.getMap().keySet().stream()
             .filter(filter)
             .map(key -> createMarshalledEntry(ctx, key))
             .iterator();
-      persistenceManager.writeBatchToAllNonTxStores(iterable, mode, cmd.getFlagsBitSet());
+
+      // Ignore empty batches
+      if (iterator.hasNext()) {
+         persistenceManager.writeBatchToAllNonTxStores(() -> iterator, mode, cmd.getFlagsBitSet());
```

Comment:
I don't think this is right. `Iterable.iterator()` is supposed to return a fresh `Iterator` every time, and some store implementations may rely on that.

Reply:
Yes good point, I'll push the fix